File: standalone/mlaz/one.py
# **************************************************************************** #
#                                                                              #
#                                                                              #
#    one.py                                                                    #
#                                                                              #
#    By: ngoguey <ngoguey@airware.com>                                         #
#                                                                              #
#    Created: 2017/04/14 18:07:57 by ngoguey                                   #
#    Updated: 2017/04/16 16:54:38 by ngoguey                                   #
#                                                                              #
# **************************************************************************** #

import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.cross_validation import train_test_split
from sklearn.preprocessing import Imputer, LabelEncoder, OneHotEncoder, StandardScaler

np.set_printoptions(linewidth=200, precision=3, suppress=True)

def dump(str_):
    v = eval(str_)
    print("******************************".format())
    s = "******** {}: {}\n{}".format(str_, type(v), v)
    s = s.replace('\n', '\n******** ')
    print(s)
    print("******************************".format())

# LOAD DATASET ************************************************************** **
dataset = pd.read_csv('one_data.csv')
dump('dataset')

x = dataset.iloc[:, 0:-1].values
dump('x')
y = dataset.iloc[:, -1].values
dump('y')

# HANDLE MISSING VALUES ***************************************************** **
print("fitting imputer to missing data".format())
imp = Imputer(missing_values='NaN', strategy='mean', axis=0)
imp = imp.fit(x[:, [1, 2]])
x[:, [1, 2]] = imp.transform(x[:, [1, 2]])
dump('x')

# CONVERT CATEGORICAL VARIABLES TO DIGITS *********************************** **
print("converting categorical variable to enum".format())
enc_x = LabelEncoder()
x[:, 0] = enc_x.fit_transform(x[:, 0])
dump('x')

print("converting enum to one-hot scheme".format())
ohenc_x = OneHotEncoder(categorical_features=[0])
dump('ohenc_x')
x = ohenc_x.fit_transform(x).toarray()
dump('x')

enc_y = LabelEncoder()
y = enc_y.fit_transform(y)
dump('y')

# TEST / TRAINING PARTITION ************************************************* **
x_train, x_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42)
dump('x_train')
dump('x_test')
dump('y_train')
dump('y_test')

# FEATURE SCALING *********************************************************** **
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)
dump('x_train')
dump('x_test')
dump('y_train')
dump('y_test')

File: standalone/mlaz/one.R
##****************************************************************************##
##                                                                            ##
##                                                                            ##
##   one.R                                                                    ##
##                                                                            ##
##   By: ngoguey <ngoguey@airware.com>                                        ##
##                                                                            ##
##   Created: 2017/04/14 18:08:01 by ngoguey                                  ##
##   Updated: 2017/05/08 13:08:19 by ngoguey                                  ##
##                                                                            ##
##****************************************************************************##

## install.packages('caTools', repos = "http://cran.us.r-project.org")
library(caTools)

## LOAD DATASET ************************************************************* **
dataset = read.csv('one_data.csv')
print('dataset:')
print(dataset)

## HANDLE MISSING VALUES **************************************************** **
dataset$Age = ifelse(
    is.na(dataset$Age), # predicate, is value missing
    ave(dataset$Age, FUN=function(x) mean(x, na.rm=TRUE)), # if true
    dataset$Age # if false
)
dataset$Salary = ifelse(
    is.na(dataset$Salary), # predicate, is value missing
    ave(dataset$Salary, FUN=function(x) mean(x, na.rm=TRUE)), # if true
    dataset$Salary # if false
)
print('dataset:')
print(dataset)

## CONVERT CATEGORICAL VARIABLES TO DIGITS ********************************** **
dataset$Country = factor(
    dataset$Country,
    levels=c('France', 'Spain', 'Germany'),
    labels=c(1, 2, 3)
)
dataset$Purchased = factor(
    dataset$Purchased,
    levels = c('No', 'Yes'),
    labels = c(0, 1)
)
print('dataset:')
print(dataset)

## TEST / TRAINING PARTITION ************************************************ **
set.seed(42)
split_mask = sample.split(dataset, SplitRatio=0.8)
training_set = subset(dataset, split_mask)
test_set = subset(dataset, split_mask == FALSE)
print('training_set:')
print(training_set)
print('test_set:')
print(test_set)

## FEATURE SCALING ********************************************************** **
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])

print('training_set:')
print(training_set)
print('test_set:')
print(test_set)
